### PR TITLE
Add local id support

### PR DESCRIPTION
**Pull request by IvanRN**
According to upcoming v1.1 of the JSONAPI the document object
identification is a combination of three values: id, type and
lid (local id). Local id is an id that can be generated when
entity was created but not yet persisted to the DB. In the
JSONAPI it should allow client to match data which he created
and then received back from the server.

### CODE REVIEW

File: jsonapi/marshal.go [hunk 1]
```diff
@@ -444,17 +456,3 @@ func marshalStruct(data MarshalIdentifier, information ServerInformation) (*Docu
 
 	return result, nil
 }
-
-func getStructType(data interface{}) string {
-	entityName, ok := data.(EntityNamer)
```
Comment (artlo) [inline jsonapi/marshal.go:449]: I think we need to save this functionality for pluralising and using structure name as json API name

File: jsonapi/unmarshal.go [hunk 2]
```diff
@@ -10,19 +10,19 @@ import (
...
 type UnmarshalIdentifier interface {
-	SetID(string) error
+	SetID(Identifier) error
 }
 
 // The UnmarshalToOneRelations interface must be implemented to unmarshal
 // to-one relations.
 type UnmarshalToOneRelations interface {
-	SetToOneReferenceID(name, ID string) error
+	SetToOneReferenceID(name string, ID *RelationshipData) error
```
Comment (artlo) [inline jsonapi/unmarshal.go]: why is RelationshipData? if we introduce Identifier struct? It's not handy to have different types for Identifiers

File: jsonapi/unmarshal.go [hunk 3]
```diff
@@ -227,9 +224,13 @@ func setRelationshipIDs(relationships map[string]Relationship, target UnmarshalI
 }
 
 func checkType(incomingType string, target UnmarshalIdentifier) error {
-	actualType := getStructType(target)
-	if incomingType != actualType {
-		return fmt.Errorf("Type %s in JSON does not match target struct type %s", incomingType, actualType)
+	castedTarget, ok := target.(MarshalIdentifier)
```
Comment (artlo) [inline jsonapi/unmarshal.go]: if you save method getStructType, you no need to cast here

File: api.go [hunk 4]
```diff
@@ -207,23 +207,14 @@ func (api *API) addResource(prototype jsonapi.MarshalIdentifier, source interfac
...
 
 	var ptrPrototype interface{}
-	var name string
 
 	if resourceType.Kind() == reflect.Struct {
 		ptrPrototype = reflect.New(resourceType).Interface()
-		name = resourceType.Name()
 	} else {
 		ptrPrototype = reflect.ValueOf(prototype).Interface()
-		name = resourceType.Elem().Name()
```
Comment (artlo) [inline api.go:217]: save getting name from Go Struct Name if GetID doesn't return name

File: api.go [hunk 5]
```diff
@@ -207,23 +207,14 @@ func (api *API) addResource(prototype jsonapi.MarshalIdentifier, source interfac
...
 	}
 
-	// check if EntityNamer interface is implemented and use that as name
-	entityName, ok := prototype.(jsonapi.EntityNamer)
```
Comment (artlo) [inline api.go:221]: this also should be saved, it looks similar that is done in function `getStructType `

File: api_interfaces_test.go [hunk 6]
```diff
@@ -120,22 +122,22 @@ var _ = Describe("Test interface api type casting", func() {
 	})
 
 	It("FindAll returns 404 for simple CRUD", func() {
-		req, err := http.NewRequest("GET", "/v1/someDatas", nil)
```
Comment (artlo) [inline api_interfaces_test.go]: save existing behaviour for api name

File: examples/model/model_chocolate.go [hunk 7]
```diff
@@ -1,19 +1,28 @@
...
 
 // GetID to satisfy jsonapi.MarshalIdentifier interface
-func (c Chocolate) GetID() string {
-	return c.ID
+func (c Chocolate) GetID() jsonapi.Identifier {
+	return jsonapi.Identifier{ID: c.ID, LID: c.LID}
+}
+
+// GetName to satisfy jsonapi.MarshalIdentifier interface
+func (c Chocolate) GetName() string {
```
Comment (artlo) [inline examples/model/model_chocolate.go]: This method is not needed anymore

File: examples/model/model_user.go [hunk 8]
```diff
@@ -18,13 +19,19 @@ type User struct {
...
 
 // GetID to satisfy jsonapi.MarshalIdentifier interface
-func (u User) GetID() string {
-	return u.ID
+func (u User) GetID() jsonapi.Identifier {
+	return jsonapi.Identifier{ID: u.ID, LID: u.LID}
+}
+
+// GetName to satisfy jsonapi.MarshalIdentifier interface
+func (u User) GetName() string {
```
Comment (artlo) [inline examples/model/model_user.go]: This method is not needed anymore

File: jsonapi/fixtures_test.go [hunk 9]
```diff
@@ -14,8 +14,8 @@ type Magic struct {
 	ID MagicID `json:"-"`
 }
 
-func (m Magic) GetID() string {
-	return m.ID.String()
+func (m Magic) GetID() Identifier {
+	return Identifier{ID: m.ID.String(), Name: "magics"}
```
Comment (artlo) [inline jsonapi/fixtures_test.go]: Save possibility ignoring Name in Indentifier

File: jsonapi/marshal_composition_test.go [hunk 10]
```diff
@@ -25,7 +29,7 @@ var _ = Describe("Embedded struct types", func() {
 			Expect(err).To(BeNil())
 			Expect(i).To(MatchJSON(`{
 				"data": {
-					"type":"taggedPosts",
```
Comment (artlo) [inline jsonapi/marshal_composition_test.go]: keep previous behaviour

File: jsonapi/marshal_composition_test.go [hunk 11]
```diff
@@ -43,7 +47,7 @@ var _ = Describe("Embedded struct types", func() {
 	Context("When unmarshaling objects with struct composition", func() {
 		postJSON := `{
 			"data": {
-				"type": "taggedPosts",
```
Comment (artlo) [inline jsonapi/marshal_composition_test.go]: the same

File: jsonapi/marshal_different_to_many_test.go [hunk 12]
```diff
@@ -71,7 +72,7 @@ var _ = Describe("Marshalling toMany relations with the same name and different
             			]
           			}
         		},
-        		"type": "manyParents"
```
Comment (artlo) [inline jsonapi/marshal_different_to_many_test.go]: the same

File: jsonapi/marshal_enum_test.go [hunk 13]
```diff
@@ -54,24 +54,25 @@ func (s *PublishStatus) UnmarshalJSON(data []byte) error {
...
+func (e *EnumPost) SetID(ID Identifier) error {
+	e.ID = ID.ID
+	e.LID = ID.LID
 	return nil
 }
 
 var _ = Describe("Custom enum types", func() {
 	status := StatusPublished
 	statusValue := "published"
-	singleJSON := []byte(`{"data":{"id": "1", "type": "enumPosts", "attributes": {"title":"First Post","status":"published"}}}`)
```
Comment (artlo) [inline jsonapi/marshal_enum_test.go]: keep previous behaviour for API name

File: jsonapi/marshal_test.go [hunk 14]
```diff
@@ -58,7 +58,7 @@ var _ = Describe("Marshalling", func() {
 			Expect(err).To(BeNil())
 			Expect(i).To(MatchJSON(`{
 				"data": {
-					"type": "simplePosts",
```
Comment (artlo) [inline jsonapi/marshal_test.go]: keep previous behaviour

File: jsonapi/marshal_test.go [hunk 15]
```diff
@@ -98,7 +98,7 @@ var _ = Describe("Marshalling", func() {
 			Expect(i).To(MatchJSON(`{
 				"data": [
 					{
-						"type": "simplePosts",
```
Comment (artlo) [inline jsonapi/marshal_test.go]: the same

**Review by artlo — CHANGES_REQUESTED** [review]

File: api.go (hunk 4 above)
Comment (IvanRN) [reply api.go:217]: saved this behaviour (get from struct if Identifier does not have name)

File: api.go (hunk 5 above)
Comment (IvanRN) [reply api.go:221]: restored this behaviour

File: jsonapi/marshal.go (hunk 1 above)
Comment (IvanRN) [reply jsonapi/marshal.go:449]: restored this functionality

File: jsonapi/marshal_test.go (hunk 15 above)
Comment (IvanRN) [reply jsonapi/marshal_test.go]: restored

File: jsonapi/marshal_test.go (hunk 14 above)
Comment (IvanRN) [reply jsonapi/marshal_test.go]: restored

File: jsonapi/marshal_enum_test.go (hunk 13 above)
Comment (IvanRN) [reply jsonapi/marshal_enum_test.go]: restored

File: jsonapi/marshal_different_to_many_test.go (hunk 12 above)
Comment (IvanRN) [reply jsonapi/marshal_different_to_many_test.go]: restored

File: jsonapi/marshal_composition_test.go (hunk 11 above)
Comment (IvanRN) [reply jsonapi/marshal_composition_test.go]: restored

File: jsonapi/unmarshal.go (hunk 3 above)
Comment (IvanRN) [reply jsonapi/unmarshal.go]: agree, restored getStructType and removed this casting

File: jsonapi/unmarshal.go (hunk 2 above)
Comment (IvanRN) [reply jsonapi/unmarshal.go]: replaced with *Identifier for ToOne and with []Identifier for ToMany

File: api_test.go [hunk 16]
```diff
@@ -37,25 +37,27 @@ func (m *requestURLResolver) SetRequest(r http.Request) {
 
 type invalid string
 
-func (i invalid) GetID() string {
-	return "invalid"
+func (i invalid) GetID() jsonapi.Identifier {
+	return jsonapi.Identifier{ID: "invalid", LID: "invalid", Name: "invalid"}
```
Comment (artlo) [inline api_test.go]: before this struct doesn't implement method GetName, so it's not needed to return `Name` for such cases.

File: api_test.go [hunk 17]
```diff
@@ -37,25 +37,27 @@ func (m *requestURLResolver) SetRequest(r http.Request) {
...
 	Value    null.Float `json:"value"`
 	Author   *User      `json:"-"`
 	Comments []Comment  `json:"-"`
 	Bananas  []Banana   `json:"-"`
 }
 
-func (p Post) GetID() string {
-	return p.ID
+func (p Post) GetID() jsonapi.Identifier {
+	return jsonapi.Identifier{ID: p.ID, LID: p.LID, Name: "posts"}
```
Comment (artlo) [inline api_test.go]: the same

File: api_test.go [hunk 18]
```diff
@@ -187,30 +189,33 @@ func (p Post) GetReferencedStructs() []jsonapi.MarshalIdentifier {
...
 type Comment struct {
 	ID    string `json:"-"`
+	LID   string `json:"-"`
 	Value string `json:"value"`
 }
 
-func (c Comment) GetID() string {
-	return c.ID
+func (c Comment) GetID() jsonapi.Identifier {
+	return jsonapi.Identifier{ID: c.ID, LID: c.LID, Name: "comments"}
```
Comment (artlo) [inline api_test.go]: the same

File: api_test.go [hunk 19]
```diff
@@ -187,30 +189,33 @@ func (p Post) GetReferencedStructs() []jsonapi.MarshalIdentifier {
...
 	ID   string `json:"-"`
+	LID  string `json:"-"`
 	Name string `json:"name"`
 	Info string `json:"info"`
 }
 
-func (u User) GetID() string {
-	return u.ID
+func (u User) GetID() jsonapi.Identifier {
+	return jsonapi.Identifier{ID: u.ID, LID: u.LID, Name: "users"}
```
Comment (artlo) [inline api_test.go]: the same

File: api_test.go [hunk 20]
```diff
@@ -187,30 +189,33 @@ func (p Post) GetReferencedStructs() []jsonapi.MarshalIdentifier {
...
 type Banana struct {
 	ID   string `jnson:"-"`
+	LID  string `jnson:"-"`
 	Name string
 }
 
-func (b Banana) GetID() string {
-	return b.ID
+func (b Banana) GetID() jsonapi.Identifier {
+	return jsonapi.Identifier{ID: b.ID, LID: b.LID, Name: "bananas"}
```
Comment (artlo) [inline api_test.go]: the same

File: examples/model/model_user.go [hunk 21]
```diff
@@ -18,13 +19,14 @@ type User struct {
 }
 
 // GetID to satisfy jsonapi.MarshalIdentifier interface
-func (u User) GetID() string {
-	return u.ID
+func (u User) GetID() jsonapi.Identifier {
+	return jsonapi.Identifier{ID: u.ID, LID: u.LID, Name: "users"}
```
Comment (artlo) [inline examples/model/model_user.go]: the same

File: examples/model/model_chocolate.go [hunk 22]
```diff
@@ -1,19 +1,23 @@
...
+	LID   string `json:"-"`
 	Name  string `json:"name"`
 	Taste string `json:"taste"`
 }
 
 // GetID to satisfy jsonapi.MarshalIdentifier interface
-func (c Chocolate) GetID() string {
-	return c.ID
+func (c Chocolate) GetID() jsonapi.Identifier {
+	return jsonapi.Identifier{ID: c.ID, LID: c.LID, Name: "chocolates"}
```
Comment (artlo) [inline examples/model/model_chocolate.go]: the same

File: jsonapi/fixtures_test.go (hunk 9 above)
Comment (artlo) [reply jsonapi/fixtures_test.go]: Not fixed

File: jsonapi/fixtures_test.go [hunk 23]
```diff
@@ -26,22 +26,35 @@ func (m MagicID) String() string {
...
+	LID              int       `json:"-"`
 	Text             string    `json:"text"`
 	SubComments      []Comment `json:"-"`
 	SubCommentsEmpty bool      `json:"-"`
 }
 
-func (c Comment) GetID() string {
-	return fmt.Sprintf("%d", c.ID)
+func (c Comment) GetID() Identifier {
+	id := Identifier{ID: fmt.Sprintf("%d", c.ID), Name: "comments"}
```
Comment (artlo) [inline jsonapi/fixtures_test.go]: No need to define Name, since GetName was not implemented before

File: jsonapi/fixtures_test.go [hunk 24]
```diff
@@ -79,27 +93,37 @@ func (c Comment) GetReferencedStructs() []MarshalIdentifier {
...
 	ID       int    `json:"-"`
+	LID      int    `json:"-"`
 	Name     string `json:"name"`
 	Password string `json:"-"`
 }
 
-func (u User) GetID() string {
-	return fmt.Sprintf("%d", u.ID)
+func (u User) GetID() Identifier {
+	id := Identifier{ID: fmt.Sprintf("%d", u.ID), Name: "users"}
```
Comment (artlo) [inline jsonapi/fixtures_test.go]: the same

File: jsonapi/fixtures_test.go [hunk 25]
```diff
@@ -123,37 +153,65 @@ type ErrorIDPost struct {
 	Error error
 }
 
-func (s ErrorIDPost) GetID() string {
-	return ""
+func (s ErrorIDPost) GetID() Identifier {
+	return Identifier{ID: "", LID: "", Name: "error-id-posts"}
```
Comment (artlo) [inline jsonapi/fixtures_test.go]: the same

File: jsonapi/fixtures_test.go [hunk 26]
```diff
@@ -123,37 +153,65 @@ type ErrorIDPost struct {
...
 	Author        *User         `json:"-"`
 	AuthorID      sql.NullInt64 `json:"-"`
+	AuthorLID     sql.NullInt64 `json:"-"`
 	AuthorEmpty   bool          `json:"-"`
 }
 
-func (c Post) GetID() string {
-	return fmt.Sprintf("%d", c.ID)
+func (c Post) GetID() Identifier {
+	id := Identifier{ID: fmt.Sprintf("%d", c.ID), Name: "posts"}
```
Comment (artlo) [inline jsonapi/fixtures_test.go]: No need to define Name, since GetName was not implemented before

File: jsonapi/fixtures_test.go [hunk 27]
```diff
@@ -272,12 +353,17 @@ func (c *Post) SetReferencedStructs(references []UnmarshalIdentifier) error {
...
 	ID       int   `json:"-"`
+	LID      int   `json:"-"`
 	AuthorID int   `json:"-"`
 	Author   *User `json:"-"`
 }
 
-func (p AnotherPost) GetID() string {
-	return fmt.Sprintf("%d", p.ID)
+func (p AnotherPost) GetID() Identifier {
+	id := Identifier{ID: fmt.Sprintf("%d", p.ID), Name: "another-posts"}
```
Comment (artlo) [inline jsonapi/fixtures_test.go]: No need to define Name, since GetName was not implemented before

File: jsonapi/fixtures_test.go [hunk 28]
```diff
@@ -301,22 +387,24 @@ func (p AnotherPost) GetReferencedIDs() []ReferenceID {
...
 	ID    string     `json:"-"`
+	LID   string     `json:"-"`
 	Title string     `json:"title"`
 	Value zero.Float `json:"value"`
 }
 
-func (z ZeroPost) GetID() string {
-	return z.ID
+func (z ZeroPost) GetID() Identifier {
+	return Identifier{ID: z.ID, LID: z.LID, Name: "zero-posts"}
```
Comment (artlo) [inline jsonapi/fixtures_test.go]: No need to define Name, since GetName was not implemented before

File: jsonapi/fixtures_test.go [hunk 29]
```diff
@@ -361,86 +449,110 @@ func (q Question) GetReferencedStructs() []MarshalIdentifier {
...
 type Identity struct {
 	ID     int64    `json:"-"`
+	LID    int64    `json:"-"`
 	Scopes []string `json:"scopes"`
 }
 
-func (i Identity) GetID() string {
-	return fmt.Sprintf("%d", i.ID)
+func (i Identity) GetID() Identifier {
+	id := Identifier{ID: fmt.Sprintf("%d", i.ID), Name: "identities"}
```
Comment (artlo) [inline jsonapi/fixtures_test.go]: No need to define Name, since GetName was not implemented before

File: jsonapi/fixtures_test.go [hunk 30]
```diff
@@ -361,86 +449,110 @@ func (q Question) GetReferencedStructs() []MarshalIdentifier {
...
+	LID            string `json:"-"`
 	Title          string
 	Number         int64
 	UnsignedNumber uint64
 }
 
-func (n *NumberPost) SetID(ID string) error {
-	n.ID = ID
+func (n NumberPost) GetID() Identifier{
+	return Identifier{ID: n.ID, LID: n.LID, Name: "number-posts"}
```
Comment (artlo) [inline jsonapi/fixtures_test.go]: No need to define Name, since GetName was not implemented before

File: jsonapi/fixtures_test.go [hunk 31]
```diff
@@ -361,86 +449,110 @@ func (q Question) GetReferencedStructs() []MarshalIdentifier {
...
 	Likes  zero.Int    `json:"likes"`
 	Rating zero.Float  `json:"rating"`
 	IsCool zero.Bool   `json:"isCool"`
 	Today  zero.Time   `json:"today"`
 }
 
-func (s SQLNullPost) GetID() string {
-	return s.ID
+func (s SQLNullPost) GetID() Identifier {
+	return Identifier{ID: s.ID, LID: s.LID, Name: "sqlNullPosts"}
```
Comment (artlo) [inline jsonapi/fixtures_test.go:521]: No need to define Name, since GetName was not implemented before

File: jsonapi/fixtures_test.go [hunk 32]
```diff
@@ -561,51 +661,45 @@ func (n CustomMetaPost) GetCustomMeta(linkURL string) map[string]Meta {
...
 type Image struct {
 	ID    string      `json:"-"`
+	LID   string      `json:"-"`
 	Ports []ImagePort `json:"image-ports"`
 }
 
-func (i Image) GetID() string {
-	return i.ID
+func (i Image) GetID() Identifier {
+	return Identifier{ID: i.ID, LID: i.LID, Name: "images"}
```
Comment (artlo) [inline jsonapi/fixtures_test.go]: No need to define Name, since GetName was not implemented before

File: jsonapi/marshal.go [hunk 33]
```diff
@@ -446,9 +457,9 @@ func marshalStruct(data MarshalIdentifier, information ServerInformation) (*Docu
 }
 
 func getStructType(data interface{}) string {
-	entityName, ok := data.(EntityNamer)
-	if ok {
-		return entityName.GetName()
+	identifier, ok := data.(MarshalIdentifier)
```
Comment (artlo) [inline jsonapi/marshal.go:460]: I think that getStructType should accept MarshalIdentifier as argument, so you no need to cast type here

**Review by artlo — CHANGES_REQUESTED** [review]

File: jsonapi/marshal.go (hunk 33 above)
Comment (IvanRN) [reply jsonapi/marshal.go:460]: I need to perform casting to MarshalIdentifier in getStructType or in the checkType. Should I return to double casting in the checkType?

File: jsonapi/fixtures_test.go (hunk 32 above)
Comment (IvanRN) [reply jsonapi/fixtures_test.go]: removed

File: jsonapi/fixtures_test.go (hunk 31 above)
Comment (IvanRN) [reply jsonapi/fixtures_test.go:521]: removed in other cases when GetName was not implemented initially, here (SQLNulPost) it actually was here before my changes

File: jsonapi/fixtures_test.go (hunk 30 above)
Comment (IvanRN) [reply jsonapi/fixtures_test.go]: removed

File: jsonapi/fixtures_test.go (hunk 29 above)
Comment (IvanRN) [reply jsonapi/fixtures_test.go]: removed

File: api_test.go (hunk 20 above)
Comment (IvanRN) [reply api_test.go]: removed

File: examples/model/model_user.go (hunk 21 above)
Comment (IvanRN) [reply examples/model/model_user.go]: removed

File: examples/model/model_chocolate.go (hunk 22 above)
Comment (IvanRN) [reply examples/model/model_chocolate.go]: removed

File: jsonapi/fixtures_test.go (hunk 23 above)
Comment (IvanRN) [reply jsonapi/fixtures_test.go]: removed

File: jsonapi/fixtures_test.go (hunk 24 above)
Comment (IvanRN) [reply jsonapi/fixtures_test.go]: removed

File: jsonapi/fixtures_test.go (hunk 25 above)
Comment (IvanRN) [reply jsonapi/fixtures_test.go]: removed

File: jsonapi/fixtures_test.go (hunk 26 above)
Comment (IvanRN) [reply jsonapi/fixtures_test.go]: removed

File: jsonapi/fixtures_test.go (hunk 27 above)
Comment (IvanRN) [reply jsonapi/fixtures_test.go]: removed

File: jsonapi/fixtures_test.go (hunk 28 above)
Comment (IvanRN) [reply jsonapi/fixtures_test.go]: removed

File: api_test.go (hunk 16 above)
Comment (IvanRN) [reply api_test.go]: removed

File: api_test.go (hunk 17 above)
Comment (IvanRN) [reply api_test.go]: removed

File: api_test.go (hunk 19 above)
Comment (IvanRN) [reply api_test.go]: removed

File: api_test.go (hunk 18 above)
Comment (IvanRN) [reply api_test.go]: removed

**Review by artlo — APPROVED** [review]

**Pull request merged**